docs/eventsV2.md

```diff
@@ -0,0 +1,268 @@
+  including `incomplete` or a type error)
+- errors during execution cause events to be retried 5 minutes later (at most twice,
+  after which they are dropped)
+
+The queue also has the following accidental features:
+
+- events do not have an execution time-limit
+- events may be cancelled if the machine they are running on turns off. This should
+  only happen to events that run longer than 28 seconds.
+
+### Design of the queue
+
+The Queue is made of 3 parts:
+
+- `events_v2` table: the backing store and source of truth for all events
+- `QueueWorker`: workers that execute events (also see EventQueueV2.fs)
+- `Google PubSub` scheduler: sends notifications to workers to potentially run an Event
+
+New definitions:
+
+- `notification`: a message containing an event ID, used to tell a queueworker which
+  event to fetch
+- `PubSub subscription`: what we read notifications from. Technically, we emit
+  notifications to a `PubSub Topic`, but the distinction isn't important.
+
+**Important**: The scheduling of events is handled by a Pub/Sub subscription, which
+**only** handles notifying the worker to try running an event. The worker may know
+that it should not run it, for example if the worker is paused. The presence of an
+Message in the PubSub subscription does not mean an actual event will be run. Nor an
+event to be run imply that there must be a
+
+When an event is first `emit`ted, we save it in the `events_v2` table, and add a
+**Notification** in PubSub. The QueueWorkers pull messages from PubSub when they
+have capacity. They then fetch the event from the `events_v2`, and perform some logic
+to verify it should be run (eg, is this handler paused, is something else already
+running this events, etc), possibly updating the stored event. It may decide to put
+the notification back into PubSub if it's not time to run it yet, or it may decide
+that it should not be run and to drop the PubSub notification. Completed events are
+deleted.
+
+### Design decisions
+
+Why use PubSub?
+
+- we spent over 50% of our DB managing worker scheduling
+
+Why not use PubSub for everything?
+
+- PubSub deletes messages after a week, so we'd have to implement extra (and difficult logic) to handle pausing
+- Similarly, emiting events in the far future would be harder to implement
+- PubSub does not support pausing
+- Queues can be paused and unpaused in quick succession by the user, meaning that
+  there can be multiple PubSub items for the same event in the queue.
+- PubSub cannot be used to count the length of the queue, and that's a really cool feature
+
+## DB schema
+
+### Events
+
+The `events_v2` table has event data
+
+- `canvas_id`, `module`, `name`, `modifier` together specify the exact handler on
+  which canvas to execute when this event is processed.
+- `value` is the emitted value (`DObj`) that was emitted to the handler
+
+It also holds scheduling metadata for an event:
+
+- `id` an id for the event. Not the same any PubSub id.
```

Comment:
> Not the same any PubSub id.
- not the same as a PubSub id?
- no two ids are the same?
- something else?

Reply:
Not the same as a PubSub id. Will update.